service/history/historyEngine.go

```diff
@@ -1728,11 +1696,48 @@ Update_History_Loop:
 
 		if updateErr != nil {
 			if updateErr == ErrConflict {
-				e.metricsClient.IncCounter(metrics.HistoryRespondDecisionTaskCompletedScope,
-					metrics.ConcurrencyUpdateFailureCounter)
+				e.metricsClient.IncCounter(metrics.HistoryRespondDecisionTaskCompletedScope, metrics.ConcurrencyUpdateFailureCounter)
 				continue Update_History_Loop
 			}
 
+			// if updateErr resulted in TransactionSizeLimitError then fail workflow
+			switch updateErr.(type) {
+			case *persistence.TransactionSizeLimitError:
+				// must reload mutable state because the first call to updateWorkflowExecutionWithContext or continueAsNewWorkflowExecution
+				// clears mutable state if error is returned
+				msBuilder, err = context.loadWorkflowExecution()
+				if err != nil {
+					return nil, err
+				}
+				attributes := &workflow.FailWorkflowExecutionDecisionAttributes{
+					Reason:  common.StringPtr(common.FailureReasonTransactionSizeExceedsLimit),
+					Details: []byte(updateErr.Error()),
+				}
+				if evt := msBuilder.AddFailWorkflowEvent(completedID, attributes); evt == nil {
```

Comment:
Probably we should use "Force Terminate" pattern and complete the workflow with Terminate event.  We do the same thing when workflow breaches overall history size and event limit.